### PR TITLE
Add event handling to the reflectometry (polref) GUI

### LINKED ISSUE #18176: Update Reflectometry GUI documentation before release 3.9
This issue is to update the Reflectometry GUI documentation before the next release. Update images, text, etc.

### LINKED ISSUE #18227: Event mode filtering: Custom slicing
This issue is the first step towards the implementation of event handling in the Reflectometry (Polref) interface.

This issue is to implement ~~uniform~~ custom (which is easier as a first step) slicing, as illustrated in this design document sent by Max:

eventhandling

Note that slicing applies to runs only, it does not apply to transmission runs.

Steps to work on this issue:

- Run Max's script to understand what it does
- Add `Uniform slicing` to tab. Radio buttons should be exclusive. Note that there should be two groups, as in tabs `Runs` and `Settings`.
- Implement the functionality: I think we only need to re-implement the `process()` method. We may find other methods defined in `GenericDataProcessorPresenter` useful, if so, we can just use them.
- We have to test as many things as possible.

### PR DESCRIPTION
**Pull request by raquelalvarezbanos**
First step to implement event handling in the Reflectometry (Polref) interface.

**To test:**

- If tested on Windows, this needs to be tested in Release mode.
- Open Interfaces > Reflectometry > Polref. There should be an extra tab called `Event Handling` (with two groups).
- Load `INTER_NR_test2.tbl` (can be found in the ISIS sample dataset) and click on `Process`, you should be able to reduce the table and get the same results as before (compare results to the same process on master).
- Remove workspaces that were created in the previous step. Use the same table, but this time remove values in column `dQ/Q`. Select both groups and hit `Process`. Groups should be processed with no errors/warnings and a workspace `IvsQ_13460_13462` should have been created.
- With the same table, go to `Event Handling` and enter a numeric value, 100 for instance. Go back to `Runs` tab and click on `Process`, you should be able to reduce the runs as event data as well, this means that two workspaces called `IvsQ_13460_0_100_13462_0_100` and `IvsQ_13469_0_100_13470_0_100` should have been created.
- Open a new table (`Reflectometry` menu > `New table`). In the first and only row enter run number `38415` (make sure the selected instrument is `INTER`) and transmission run `38393`. Reduce the row and compare results with this script provided by scientists: [EventRef.zip](https://github.com/mantidproject/mantid/files/714298/EventRef.zip). `INTER` datasets:
[reflectometry_datasets.zip](https://github.com/mantidproject/mantid/files/726256/reflectometry_datasets.zip)

- Open a new table and add run `38415` in the first row. Add a new row and add run `38417`. Enter some values in column `dQ/Q`. Process the group. The first run is an event workspace but the second one is not, you should get a warning message saying that the group could not be processed as event workspace, but some workspaces should have been created in the ADS (the interface will have reduced the runs as non-event workspaces).
- Unit tests for `ReflDataProcessorPresenter` will be added in a separate issue (https://github.com/mantidproject/mantid/issues/18498).
- Documentation has been updated.

Fixes #18227.
Also fixes #18176.

Release notes: https://github.com/mantidproject/mantid/pull/18463/commits/f58a3b80390342ca308052f697233151f0b47866

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] How do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by AnthonyLim23** [comment]
works as expected, except if an empty row is present in the table then an error is thrown.

File: MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Reflectometry/QtReflEventTabView.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  /// Initialise the interface
+  void initLayout();
+
+  /// The widget
+  Ui::ReflEventTabWidget m_ui;
+  /// The presenter
+  std::unique_ptr<IReflEventTabPresenter> m_presenter;
+};
+
+} // namespace Mantid
```
Comment (AnthonyLim23) [inline MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Reflectometry/QtReflEventTabView.h]: why are these namespaces comments?

File: MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Reflectometry/ReflDataProcessorPresenter.h [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                             const std::vector<double> &startTimes,
+                             const std::vector<double> &stopTimes);
+  // Process a group of runs which are not event workspaces
+  bool processGroupAsNonEventWS(int groupID, const GroupData &group);
+  // Parse time slicing from string
+  void parseTimeSlicing(const std::string &timeSlicing,
+                        std::vector<double> &startTimes,
+                        std::vector<double> &stopTimes);
+  // Load a run as event workspace
+  bool loadEventRun(const std::string &runno);
```
Comment (AnthonyLim23) [inline MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Reflectometry/ReflDataProcessorPresenter.h]: should it not be runNo?

File: MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Reflectometry/QtReflEventTabView.h (hunk 1 above)
Comment (raquelalvarezbanos) [inline MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Reflectometry/QtReflEventTabView.h]: No reason, I probably copied-pasted them from another class.

File: MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Reflectometry/ReflDataProcessorPresenter.h (hunk 2 above)
Comment (raquelalvarezbanos) [inline MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Reflectometry/ReflDataProcessorPresenter.h]: Strictly, yes it should be `runNo`, I'll change that.

**Comment by raquelalvarezbanos** [comment]
I fixed the issue you found when a row is empty. It should be fixed now.

**Comment by AnthonyLim23** [comment]
I just checked and it fixed the problem.

**Comment by AnthonyLim23** [comment]
once build is complete :shipit:

**Pull request merged**